### PR TITLE
Run tests on multiple CPUs

### LINKED ISSUE #1406: Get to passing tests with multiple cpus
Currently failing:
- [x] multiraft (~~#1458~~)
- [x] server (~~#1456~~, ~~#1459~~)
- [x] server/status (~~#1436~~)
- [x] storage (~~#1466~~)
- [x] util/metrics

### PR DESCRIPTION
**Pull request by tamird**
Closes #1406.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,8 +35,8 @@ RUN := run
 PKG          := ./...
 TAGS         :=
 TESTS        := ".*"
-TESTTIMEOUT  := 15s
-CPUS         := 1
+TESTTIMEOUT  := 1m
+CPUS         := 2,4
```
Comment (spencerkimball) [inline Makefile]: I wonder whether it would be better to use 1,4 here since the =1 case is more different from =4 than =2 is.

**Comment by spencerkimball** [comment]
LGTM

**Comment by bdarnell** [comment]
LGTM, but I agree with Spencer that 1,4 is probably better than 2,4.

**Comment by tamird** [comment]
I'm going to include all three, because I'm pretty sure circleCI runs dual-core machines

**Comment by MBerhault** [comment]
Doesn't this run the tests once for each value of -cpu? doubling or trebling seems a bit unwise.

**Comment by bdarnell** [comment]
The time taken to run all the tests three times will add up. Even twice is pushing it, although I think there may be some value to it since it means that non-deterministic failures are less likely to slip through.

**Comment by spencerkimball** [comment]
How much more slowly did the tests run?

**Comment by MBerhault** [comment]
How about we just do that for master builds?

**Comment by tschottdorf** [comment]
Just as an FYI, we'll eventually want nightly builds. Looking at our growing suite of acceptance test that "eventually" may be soon, probably after #1115.

**Comment by tamird** [comment]
Before:
test: 01:35
testrace: 03:18

After:
test: 02:01
testrace: 07:03

So total impact on build time is ~4m30s.

**Comment by spencerkimball** [comment]
In that case, we should consider just using CPU=4 as the default and only
running a collection of different CPU values on nightly builds.

**Comment by tamird** [comment]
Let me circle back on that. I'm thinking we'll actually want `1,2` or `2,2` because:
- I'm pretty sure circle machines have two logical CPUs
- We want to run the test suite twice to make sure we don't regress with respect to tests polluting global state

**Comment by bdarnell** [comment]
I disagree with the second bullet. In most cases, we detect problems involving mutations to global state because one test will affect another within a single run. We may occasionally have problems that are not detected until we run the entire test suite twice in one process, but this is not enough to justify running every test twice on every run.

**Comment by spencerkimball** [comment]
+1 Let's not add a major additional chunk of time to the circleci runs unless it's truly useful. Trying a variety of different CPU counts in a nightly run would completely obviate the utility of this and save us an extra 4.5m.

**Comment by tamird** [comment]
Final tally:
`1,2`:
test: 01:44
testrace: 04:12

`2`:
test: 1:31
testrace: 03:17

TestServerNodeEventFeed and TestTxnDBLostUpdateAnomaly still occassionally fail, but we have a green build now.

**Comment by spencerkimball** [comment]
Is there an issue for TestTxnDBLostUpdateAnomaly? That needs to be tracked
down.

On Wed, Jun 24, 2015 at 11:59 PM, Tamir Duberstein <notifications@github.com

**Comment by spencerkimball** [comment]
I'm actually coming to the conclusion that the Makefile should just specify -cpu=4.

Seeing double test output when there's a problem is becoming onerous in practice. I think the chance of finding an issue with =1 or =2 when we already have =4 is small enough that doubling up the output is a waste of time and cognitive effort. Seems sufficient to include a variety of different cpu settings with nightly builds.

File: Makefile [hunk 2]
```diff
@@ -35,8 +35,8 @@ RUN := run
 PKG          := ./...
 TAGS         :=
 TESTS        := ".*"
-TESTTIMEOUT  := 15s
-CPUS         := 1
+TESTTIMEOUT  := 1m
```
Comment (spencerkimball) [inline Makefile]: Let's not make this so large; it's disconcerting. This change would remove the incentive to find out why things are slowing.

File: storage/id_alloc_test.go [hunk 3]
```diff
@@ -176,7 +176,7 @@ func TestAllocateErrorAndRecovery(t *testing.T) {
 		}()
 	}
 	// Make sure no allocation returns.
-	time.Sleep(10 * time.Millisecond)
+	time.Sleep(250 * time.Millisecond)
```
Comment (spencerkimball) [inline storage/id_alloc_test.go]: Why should this be necessary? Waiting a shorter amount of time shouldn't trouble a slow machine. This just adds a quarter second to every test run--we should resist making such changes.

**Comment by spencerkimball** [comment]
LGTM

File: storage/client_raft_test.go [hunk 4]
```diff
@@ -382,7 +389,7 @@ func TestFailedReplicaChange(t *testing.T) {
 
 	// The pending config change flag was cleared, so a subsequent attempt
 	// can succeed.
-	storage.TestingCommandFilter = nil
+	runFilter.Store(false)
```
Comment (tschottdorf) [inline storage/client_raft_test.go]: @kkaneda just merged something new that uses this filter, so it'll likely need an update.

File: storage/command_queue_test.go [hunk 5]
```diff
@@ -166,8 +166,8 @@ func TestCommandQueueClear(t *testing.T) {
 	// Clear the queue and verify both commands are signaled.
 	cq.Clear()
 
-	if !testCmdDone(cmdDone1, 1*time.Millisecond) ||
-		!testCmdDone(cmdDone2, 1*time.Millisecond) {
+	if !testCmdDone(cmdDone1, 100*time.Millisecond) ||
+		!testCmdDone(cmdDone2, 100*time.Millisecond) {
```
Comment (tschottdorf) [inline storage/command_queue_test.go]: does this waste 200ms no matter what?

**Comment by tschottdorf** [comment]
LGTM. I would also prefer not adding a large amount of time to our daytime runs. Going nightly isn't a lot of work.

File: storage/id_alloc_test.go (hunk 3 above)
Comment (tamird) [inline storage/id_alloc_test.go]: Good point. Fixed with a better synchronization solution.

File: storage/command_queue_test.go (hunk 5 above)
Comment (tamird) [inline storage/command_queue_test.go]: no, `cmdDone{1,2}` and `testCmdDone` is implemented with a `select`

File: storage/client_raft_test.go (hunk 4 above)
Comment (tamird) [inline storage/client_raft_test.go]: nah, only this change is only necessary when `TestingCommandFilter` is being turned off in the middle of a test (and causing a race)

Comment (kkaneda) [inline storage/client_raft_test.go]: Ah, thanks! Will rebase after this PR is merged.

File: Makefile (hunk 2 above)
Comment (tamird) [inline Makefile]: OK, ratcheted down to 30s. It'd be great to get a list of slow tests, but that's not something you get from the standard test runner AFAIK.

Comment (tschottdorf) [inline Makefile]: `make test TESTFLAGS=-v | grep ' PASS'`?

**Comment by bdarnell** [comment]
All the changes LGTM, but as we just discussed we should hold off on actually changing the default for `CPUS` until the remaining tests have stabilized.

**Comment by tamird** [comment]
Remaining failures:
- [ ] TestHeartbeatMultipleGroupsJointLeader ([4062](https://circleci.com/gh/cockroachdb/cockroach/4062), [4074](https://circleci.com/gh/cockroachdb/cockroach/4074))
- [ ] TestTxnDBLostUpdateAnomaly ([4032](https://circleci.com/gh/cockroachdb/cockroach/4032), [4045](https://circleci.com/gh/cockroachdb/cockroach/4045))
- [ ] TestStoreRaftIDAllocation ([4068](https://circleci.com/gh/cockroachdb/cockroach/4068))
- [ ] TestServerNodeEventFeed ([4033](https://circleci.com/gh/cockroachdb/cockroach/4033))
- [ ] TestMultiStoreEventFeed ([4043](https://circleci.com/gh/cockroachdb/cockroach/4043), [4062](https://circleci.com/gh/cockroachdb/cockroach/4062))

File: storage/client_raft_test.go (hunk 4 above)
Comment (kkaneda) [inline storage/client_raft_test.go]: Hmm, I've changed the test to use a `defer, but still got a race error:

https://circle-artifacts.com/gh/cockroachdb/cockroach/4137/artifacts/0/tmp/circle-artifacts.eLAGyKb/excerpt.txt

```
WARNING: DATA RACE
Read by goroutine 49:
  github.com/cockroachdb/cockroach/storage.(*Range).executeCmd()
      /go/src/github.com/cockroachdb/cockroach/storage/range_command.go:53 +0x4d9
  github.com/cockroachdb/cockroach/storage.(*Range).applyRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:888 +0xa0d
  github.com/cockroachdb/cockroach/storage.(*Range).processRaftCommand()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:776 +0x372
  github.com/cockroachdb/cockroach/storage.funcÂ·043()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:1492 +0xd5d
  github.com/cockroachdb/cockroach/util.funcÂ·001()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:75 +0x7f

Previous write by goroutine 205:
  github.com/cockroachdb/cockroach/storage_test.funcÂ·032()
      /go/src/github.com/cockroachdb/cockroach/storage/client_range_test.go:301 +0x38
  github.com/cockroachdb/cockroach/storage_test.TestTxnPutOutOfOrder()
      /go/src/github.com/cockroachdb/cockroach/storage/client_range_test.go:311 +0x163e
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133
```

Do I need to make the test wait all activities finish?

**Comment by tamird** [comment]
I've added a few commits and I think this is ready to merge. PTAL

**Comment by tschottdorf** [comment]
You're right that the `LostUpdateAnomaly` error is expected on slow machines (as discussed). We limit to three retries, but there's no reason that this is enough. Simply set the limit higher.

**Comment by tschottdorf** [comment]
Happy to see this merged, unless there's a test failure open we should definitely fix before.

**Comment by tamird** [comment]
@tschottdorf addressed your comments

**Comment by tschottdorf** [comment]
looks like this got merged with three retries in `txn_correctness_test.go`, was this on purpose? I get getting the anomaly test failures because of that on my branch.

**Comment by spencerkimball** [comment]
Setting the limit higher means those anomaly tests will take a lot longer
to run--we rely on the number of retries being exceeded when forward
progress cannot be made.

On Mon, Jun 29, 2015 at 4:15 AM, Tobias Schottdorf <notifications@github.com

**Comment by tschottdorf** [comment]
I see - you're running all possible histories, and all that don't error out are taken into account?

**Comment by spencerkimball** [comment]
If you error out, it assumes it's because there's no way to push a
particular txn (priority too high), so after erroring out, that txn will
abort and restart.

**Comment by tschottdorf** [comment]
Ok, so it looks like we need to change the retry options selectively. For those tests with `expSuccess=true` we set infinite retries, for those supposed to fail we stick with what we have. Am I missing a caveat?

**Comment by spencerkimball** [comment]
No, that's not a bad solution to this non-deterministic load-related
failure.

**Pull request merged**